cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -373,12 +388,75 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
     store.deleteSchedule(workflowId, SchedulableProgramType.WORKFLOW, scheduleName);
   }
 
-  private void startWorkerAdapter(Id.Namespace namespace, AdapterSpecification adapterSpec) {
-    // TODO: implement
+  private Id.Program getWorkerId(Id.Namespace namespace, AdapterSpecification adapterSpec) throws NotFoundException {
+    Id.Application appId = Id.Application.from(namespace, adapterSpec.getTemplate());
+    ApplicationSpecification appSpec = store.getApplication(appId);
+    if (appSpec == null || appSpec.getWorkers().size() != 1) {
+      throw new NotFoundException(appId);
+    }
+    String workflowName = Iterables.getFirst(appSpec.getWorkers().keySet(), null);
+    return Id.Program.from(namespace.getId(), adapterSpec.getTemplate(), ProgramType.WORKER, workflowName);
   }
 
-  private void stopWorkerAdapter(Id.Namespace namespace, AdapterSpecification adapterSpec) {
-    // TODO: implement
+  private void startWorkerAdapter(Id.Namespace namespace, AdapterSpecification adapterSpec) throws NotFoundException {
+    final Id.Adapter adapterId = Id.Adapter.from(namespace.getId(), adapterSpec.getName());
+    final Id.Program workerId = getWorkerId(namespace, adapterSpec);
+    String workerName = workerId.getId();
+    try {
+      Map<String, String> sysArgs = resolver.getSystemProperties(workerId, ProgramType.WORKER);
+      Map<String, String> userArgs = resolver.getUserProperties(workerId, ProgramType.WORKER);
+      // Override resolved preferences with adapter worker spec properties.
+      userArgs.putAll(adapterSpec.getRuntimeArgs());
+      store.setWorkerInstances(workerId, adapterSpec.getInstances());
+      ProgramRuntimeService.RuntimeInfo runtimeInfo = lifecycleService.start(workerId, ProgramType.WORKER,
+                                                                             sysArgs, userArgs, false);
+      adapterStore.setRunId(adapterId, runtimeInfo.getController().getRunId());
```

Comment:
not introduced by you, but we can have races where requests to start the same program are made concurrently and then things break.  should probably have some sort of check and set logic on program status before we do anything.  Can you open a jira and add a todo somewhere?